### PR TITLE
5pm-a search by instructor

### LINKED ISSUE #32: Search By instructor (over multiple quarters)
# User Story

As a user, I can enter an instructor name and a start and end quarter,  and see a list of information about different class that this instructor taught, so that I can see if one of my favorite professor is teaching a class this quarter.

# Acceptance Criteria

- [x] There is a menu option that allows me to select "Search by instructor"
- [x] The "Search by instructor" menu option takes me to a page where I can enter instructor name, start and end quarter.
- [x] When I click the "submit" button (or search or whatever), I see a list of course that the professor has taught, along with the normal information about each course.

Note: we can reuse the component for listing courses, so deciding what gets displayed for each course need not be a part of this story; whatever the component provides is fine.

# Implementation Todos

Front end:
- [x] Add the item to the navigation bar to route to this page.
- [x] Add the new page to `App.js` (this is standard for adding a new page in our React apps).
- [x] Implement a component for the form where you enter instructor name, and select start/end quarter.
- [x] Initially, you can display some fake course data until the backend route is completed.
- [x] Display real data that you get from a backend route.  Decide on the api route name.

Back end:

- [x] Add a backend route for the front end to retrieve this data
- [x] There may also some other work involved in interacting with the MongoDB database... this is TBD.. consult the staff.

Testing:

- [x] Front end tests pass and there is adequate coverage
- [x] Back end tests pass and there is adequate coverage

### PR DESCRIPTION
**Pull request by lpalomera**
This pull request closes #32. This branch adds a page and form where a user can enter an instructor, and start/endquarter and receive the respective information. The branch also adds a new controller for the back end to interact with the archived course repository (as well as adds respective testing for new files added in the front and back end).

# Examples

Single Quarter:

course-search-by-instructor

Multiple Quarters:

course-search-by-instructor-multiple-quarters

# Known bug: CMPSC 596 courses give incorrect results.

There is a known bug with this feature that is being left to W21 to resolve: namely, there is a certain kind of course (CMPSC 596 is an example) where a large number of instructors are associated with the course.   CMPSC 596 is a course where instructors supervise the work of individual graduate students working on research.  

Steps to Reproduce: Search for an instructor that teaches a CMPSC 596 course, e.g. Conrad

Observed: Every instructor's CMPSC 596 section is included in the search results.

Desired: Only the searched for instructor's CMPSC 596 results are returned.

Note: There may be a temptation to solve this by writing special case for for CMPSC 596, but that would not be a solid approach; there are other such courses across campus.  We should look into what the pattern is in the underlying data (the JSON representation) and how to computationally distinguish this case from the case of regular courses.

### CODE REVIEW

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    return (
+        <Form onSubmit={handleSubmit}>
+            <Form.Group controlId="InstructorSearch.StartController">
+                <Form.Label>Start Quarter</Form.Label>
+                <Form.Control as="select" onChange={handleStartQuarterOnChange} value={startQuarter}  >
+                    <option value="20211">W21</option>
+                    <option value="20204">F20</option>
+                    <option value="20201">W20</option>
+                    <option value="20202">S20</option>
+                    <option value="20203">M20</option>
```
Comment (kevbbn) [inline javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js:48]: I would order these in chronological order, I think you just need to put W21 on the bottom

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                </Form.Control>
+            </Form.Group>
+            <Form.Group controlId="Instructor.EndController">
+                <Form.Label>End Quarter</Form.Label>
+                <Form.Control as="select" onChange={handleEndQuarterOnChange} value={endQuarter}  >
+                    <option value="20211">W21</option>
+                    <option value="20204">F20</option>
+                    <option value="20201">W20</option>
+                    <option value="20202">S20</option>
+                    <option value="20203">M20</option>
```
Comment (kevbbn) [inline javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js:59]: I would order these in chronological order, I think you just need to put W21 on the bottom

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                </Form.Control>
+            </Form.Group>
+            <Form.Group controlId="InstructorSearch.Instructor">
+                <Form.Label>Instructor</Form.Label>
+                <Form.Control type="text" onChange={handleInstructorOnChange} value={instructorText}>
+
```
Comment (kevbbn) [inline javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js:65]: we can remove this newline

File: javascript/src/main/pages/History/Instructor.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    };
+
+    const [courseJSON, setCourseJSON] = useState(initialCourseJSON);
+
+
+    // const courseJSON = '{"course" : "cs156"}';
+    return (
+        <Jumbotron>
+            <div className="text-left">
+                <h5>Search Archived Course Data from MongoDB</h5>
```
Comment (kevbbn) [inline javascript/src/main/pages/History/Instructor.js:36]: Add a header here that says you are searching by instructor 

<h2>Search Archived Course Data from MongoDB</h2>
<h5>Search Instructor Through Various Quarters</h5>

heres the code to match search by ge and search by coursename

File: src/main/java/edu/ucsb/courses/controllers/HistorySearchInstructQtrController.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    ArchivedCourseRepository archivedCourseRepository;
+
+
+    @GetMapping(value = "/instructorsearch", produces = "application/json")
+    public ResponseEntity<String> instructorsearch(
+        @RequestParam String startQtr, 
+        @RequestParam String endQtr, 
+        //@RequestParam String subjectArea,
+        //@RequestParam String courseNumber,
+        //@RequestParam String courseSuf
```
Comment (kevbbn) [inline src/main/java/edu/ucsb/courses/controllers/HistorySearchInstructQtrController.java:39]: not the biggest of deals but all of this can be removed

**Review by kevbbn — CHANGES_REQUESTED** [review]
I'd add the header that says "Search By Instructor" and order the quarters, then we should be good!

**Review by kevbbn — APPROVED** [review]
LGTM

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    const [startQuarter, setStartQuarter] = useState("20204");
+    const [endQuarter, setEndQuarter] = useState("20204");
+    const [instructorText, setInstructor]=useState("");
+    
+
+    const handleSubmit = (event) => {
+        //console.log(event);
+        event.preventDefault();
+        fetchJSON(event, { startQuarter, endQuarter, instructorText}).then((courseJSON) => {
```
Comment (KobeShav) [inline javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js:17]: Delete console.log

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            setCourseJSON(courseJSON);
+        });
+    };
+
+    // FALL = 4, F
+    // WINTER = 1, W
+    // SPRING = 2, S
+    // SUMMER = 3, M
+    const handleStartQuarterOnChange = (event) => {
+        setStartQuarter(event.target.value);
```
Comment (KobeShav) [inline javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js:27]: Don't need these comments anymore

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    const handleEndQuarterOnChange = (event) => {
+        setEndQuarter(event.target.value);
+    };
+
+    const handleInstructorOnChange = (event) => {
+        setInstructor(event.target.value);
+    };
+
+    // Note:
```
Comment (KobeShav) [inline javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js:38]: Same, extra comment

File: javascript/src/main/pages/History/Instructor.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import {  fetchInstructorHistoryNameQtrJSON } from "main/services/courseSearches";
+
+const Instructor = () => {
+
+    // every function that starts with "use" is a hook
+    // e.g. useState, useSWR, useAuth0
+
+    // courseJSON is the variable for the state
+    // setCourseJSON is the setter
+    // the parameter to useState is the initial value of the state
```
Comment (KobeShav) [inline javascript/src/main/pages/History/Instructor.js:21]: I was told to remove these so you should too

File: javascript/src/main/pages/Home/Home.js [hunk 10]
```diff
@@ -6,6 +6,7 @@ import { Redirect } from "react-router-dom";
 import BasicCourseSearchForm from "main/components/BasicCourseSearch/BasicCourseSearchForm";
 import JSONPrettyCard from "main/components/Utilities/JSONPrettyCard";
 import { fetchBasicCourseJSON } from "main/services/courseSearches";
+//import {fetchInstructorHistoryNameQtrJSON} from "main/services/courseSearches";
```
Comment (KobeShav) [inline javascript/src/main/pages/Home/Home.js:9]: Instead of a comment, it's probably cleaner to just delete it

**Review by KobeShav — COMMENTED** [review]
Just fix some white space and extra comments and we're good to go!

File: javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js (hunk 7 above)
Comment (pconrad) [reply javascript/src/main/components/BasicCourseSearch/CourseSearchFormInstructor.js:27]: Normally, I'd agree with deleting comments (especially if it's commented out code), but I think in this case, these might be ok to keep.

File: javascript/src/main/pages/History/Instructor.js (hunk 9 above)
Comment (pconrad) [reply javascript/src/main/pages/History/Instructor.js:21]: lol... yeah, these were here mostly to help folks when we were just starting out.   I agree that these can probably go.

File: javascript/src/main/pages/Home/Home.js (hunk 10 above)
Comment (pconrad) [reply javascript/src/main/pages/Home/Home.js:9]: Yes, in this case, definitely

**Review by pconrad — APPROVED** [review]
LGTM

**Comment by pconrad** [comment]
The code review went fine.  But, this appears to have some bugs.

One bug is that if you search for an instructor, and their name appears in one of the CMPSC 596 sections, then ALL of the CMPSC 596 sections show up.    


See demos below.

course-search-pr-79-after

**Comment by andrewhlu** [comment]
@pconrad Looks like your GIF didn't upload properly, I'm unable to see the full demo. Can you try uploading it again?

**Comment by phtcon** [comment]
Here's the missing after gif
course-search-pr-79-after

course-search-pr-79-after-p2

I see a message on Slack that this is now fixed.. I'm going to check it out.

**Comment by KobeShav** [comment]
Search by instructor now working to return JSON objects where the instructor field matches the input (using regular expressions). One known bug is that any class (such as CMPSC 596) which has a match for 1 professor will return the object for all of the professors, resulting in a bloated table.

**Review by pconrad — APPROVED** [review]
LGTM!

**Comment by pconrad** [comment]
Functionality tested, and looks good.   The known bug has been filed as something to fix in W21.

In W21, we can also add the quarter to the output; that's helpful since we are displaying output for multiple quarters.

**Comment by pconrad** [comment]
Circling back to this.. a deployment is pending here: https://ucsb-courses-search-f20-qa.herokuapp.com/

When it completes, I'll do a final QA on this and then I think it's mergeable.

**Comment by pconrad** [comment]
Has a known bug, but good enough for a starting point and we can iterate on this in W21

**Pull request merged**